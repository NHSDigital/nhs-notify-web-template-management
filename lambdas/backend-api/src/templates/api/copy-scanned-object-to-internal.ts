import { z } from 'zod';
import type { TemplateFileScannedEventDetail } from 'nhs-notify-web-template-management-utils';
import type { LetterUploadRepository } from '../infra/letter-upload-repository';

type CopyScannedObjectToInternalLambdaInput = {
  detail: TemplateFileScannedEventDetail;
};

const $CopyScannedObjectToInternalLambdaInput: z.ZodType<CopyScannedObjectToInternalLambdaInput> =
  z.object({
    detail: z.object({
      template: z.object({
        id: z.string(),
        owner: z.string(),
      }),
      versionId: z.string(),
      fileType: z.enum(['pdf-template', 'test-data']),
      virusScanStatus: z.enum(['PASSED']),
    }),
  });

export const createHandler =
  ({
    letterUploadRepository,
  }: {
    letterUploadRepository: LetterUploadRepository;
  }) =>
  async (event: unknown) => {
    const {
      detail: { template, versionId, fileType },
    } = $CopyScannedObjectToInternalLambdaInput.parse(event);

    await letterUploadRepository.copyFromQuarantineToInternal(
      template,
      fileType,
      versionId
    );
  };
