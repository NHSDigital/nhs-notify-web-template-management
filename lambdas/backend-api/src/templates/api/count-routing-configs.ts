import type { APIGatewayProxyHandler } from 'aws-lambda';
import { apiFailure, apiSuccess } from './responses';
import type { RoutingConfigClient } from '../app/routing-config-client';
import { logger } from 'nhs-notify-web-template-management-utils/logger';

export function createHandler({
  routingConfigClient,
}: {
  routingConfigClient: RoutingConfigClient;
}): APIGatewayProxyHandler {
  return async function handler(event) {
    const { user: userId, clientId } = event.requestContext.authorizer ?? {};

    if (!clientId || !userId) {
      return apiFailure(400, 'Invalid request');
    }

    const user = { userId, clientId };

    const log = logger.child(user);

    const { data, error } = await routingConfigClient.countRoutingConfigs(
      user,
      event.queryStringParameters
    );

    if (error) {
      log
        .child(error.errorMeta)
        .error('Failed to count routing configs', error.actualError);

      return apiFailure(
        error.errorMeta.code,
        error.errorMeta.description,
        error.errorMeta.details
      );
    }

    return apiSuccess(200, data);
  };
}
