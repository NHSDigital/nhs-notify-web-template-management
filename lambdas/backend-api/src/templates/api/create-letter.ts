import type { APIGatewayProxyHandler } from 'aws-lambda';
import { apiFailure, apiSuccess } from './responses';
import type {
  CreateTemplate,
  ITemplateClient,
} from 'nhs-notify-backend-client';
import { getLetterUploadParts } from '../app/get-letter-upload-parts';

export function createHandler({
  templateClient,
}: {
  templateClient: ITemplateClient;
}): APIGatewayProxyHandler {
  return async function (event) {
    const user = event.requestContext.authorizer?.user;

    if (!user) {
      return apiFailure(400, 'Invalid request');
    }

    const base64body = Buffer.from(event.body ?? '', 'base64');

    const contentType =
      event.headers['Content-Type'] ?? event.headers['content-type'] ?? 'none';

    const { error: getLetterPartsError, data: letterParts } =
      getLetterUploadParts(base64body, contentType);

    if (getLetterPartsError) {
      return apiFailure(getLetterPartsError.code, getLetterPartsError.message);
    }

    const { template, pdf, csv } = letterParts;

    const { data: created, error: createTemplateError } =
      await templateClient.createLetterTemplate(
        template as CreateTemplate,
        user,
        pdf,
        csv
      );

    if (createTemplateError) {
      return apiFailure(
        createTemplateError.code,
        createTemplateError.message,
        createTemplateError.details
      );
    }

    return apiSuccess(201, created);
  };
}
