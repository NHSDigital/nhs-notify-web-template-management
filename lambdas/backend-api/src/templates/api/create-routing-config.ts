import type { APIGatewayProxyHandler } from 'aws-lambda';
import { apiFailure, apiSuccess } from './responses';
import type { RoutingConfigClient } from '../app/routing-config-client';
import { logger } from 'nhs-notify-web-template-management-utils/logger';

export function createHandler({
  routingConfigClient,
}: {
  routingConfigClient: RoutingConfigClient;
}): APIGatewayProxyHandler {
  return async function handler(event) {
    const { user: userId, clientId } = event.requestContext.authorizer ?? {};

    if (!clientId || !userId) {
      return apiFailure(400, 'Invalid request');
    }

    const payload = JSON.parse(event.body || '{}');

    const user = {
      clientId,
      userId,
    };

    const log = logger.child(user);

    const { data, error } = await routingConfigClient.createRoutingConfig(
      payload,
      user
    );

    if (error) {
      log
        .child(error.errorMeta)
        .error('Failed to create routing configuration', error.actualError);

      return apiFailure(
        error.errorMeta.code,
        error.errorMeta.description,
        error.errorMeta.details
      );
    }

    return apiSuccess(201, data);
  };
}
