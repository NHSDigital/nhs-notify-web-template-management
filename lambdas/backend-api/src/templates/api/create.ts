import type { APIGatewayProxyHandler } from 'aws-lambda';
import { TemplateClient } from '@backend-api/templates/app/template-client';
import { apiFailure, apiSuccess } from './responses';

export const handler: APIGatewayProxyHandler = async (event) => {
  const user = event.requestContext.authorizer?.user;

  const dto = JSON.parse(event.body || '{}');

  if (!user) {
    return apiFailure(400, 'Invalid request');
  }

  const client = new TemplateClient(user, false);

  const { data, error } = await client.createTemplate(dto);

  if (error) {
    return apiFailure(error.code, error.message, error.details);
  }

  return apiSuccess(201, data);
};
