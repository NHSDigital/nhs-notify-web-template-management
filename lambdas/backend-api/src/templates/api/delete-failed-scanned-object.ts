import { guardDutyEventValidator } from 'nhs-notify-web-template-management-utils';
import type { LetterFileRepository } from '../infra/letter-file-repository';

export const createHandler =
  ({ letterFileRepository }: { letterFileRepository: LetterFileRepository }) =>
  async (event: unknown) => {
    const {
      detail: {
        s3ObjectDetails: { objectKey, versionId },
      },
    } = guardDutyEventValidator('FAILED').parse(event);

    await letterFileRepository.deleteFromQuarantine(objectKey, versionId);
  };
