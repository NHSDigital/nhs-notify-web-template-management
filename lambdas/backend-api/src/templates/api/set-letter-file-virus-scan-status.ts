import { z } from 'zod';
import {
  $GuardDutyMalwareScanStatus,
  type GuardDutyMalwareScanStatus,
} from 'nhs-notify-web-template-management-utils';
import type { TemplateRepository } from '../infra';
import type { LetterUploadMetadata } from '../infra/letter-upload-repository';

type SetLetterFileVirusScanStatusLambdaInput = {
  detail: {
    object: {
      metadata: Omit<
        LetterUploadMetadata,
        'user-filename' | 'test-data-provided'
      >;
      tags: { GuardDutyMalwareScanStatus: GuardDutyMalwareScanStatus };
    };
  };
};

const $SetLetterFileVirusScanStatusLambdaInput: z.ZodType<SetLetterFileVirusScanStatusLambdaInput> =
  z.object({
    detail: z.object({
      object: z.object({
        metadata: z.object({
          owner: z.string(),
          'template-id': z.string(),
          'version-id': z.string(),
          'file-type': z.enum(['pdf-template', 'test-data']),
        }),
        tags: z.object({
          GuardDutyMalwareScanStatus: $GuardDutyMalwareScanStatus,
        }),
      }),
    }),
  });

export const createHandler =
  ({ templateRepository }: { templateRepository: TemplateRepository }) =>
  async (event: unknown) => {
    const {
      detail: {
        object: { metadata, tags },
      },
    } = $SetLetterFileVirusScanStatusLambdaInput.parse(event);

    await templateRepository.setLetterFileVirusScanStatus(
      { owner: metadata.owner, id: metadata['template-id'] },
      metadata['file-type'] === 'pdf-template' ? 'pdfTemplate' : 'testDataCsv',
      metadata['version-id'],
      tags.GuardDutyMalwareScanStatus === 'NO_THREATS_FOUND'
        ? 'PASSED'
        : 'FAILED'
    );
  };
