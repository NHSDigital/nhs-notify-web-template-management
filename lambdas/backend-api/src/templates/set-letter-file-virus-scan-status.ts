import { createHandler } from './api/set-letter-file-virus-scan-status';
import { createContainer } from './container';

export const handler = createHandler(createContainer());
