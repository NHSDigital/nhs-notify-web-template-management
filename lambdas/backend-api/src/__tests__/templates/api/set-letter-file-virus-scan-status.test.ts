import { mock } from 'jest-mock-extended';
import { makeGuardDutyMalwareScanResultNotificationEvent } from 'nhs-notify-web-template-management-test-helper-utils';
import { $GuardDutyMalwareScanStatusFailed } from 'nhs-notify-web-template-management-utils';
import type {
  TemplateRepository,
  LetterUploadRepository,
} from '../../../templates/infra';
import { createHandler } from '../../../templates/api/set-letter-file-virus-scan-status';

function setup() {
  const mocks = {
    templateRepository: mock<TemplateRepository>(),
    letterUploadRepository: mock<LetterUploadRepository>(),
  };
  const handler = createHandler(mocks);
  return { handler, mocks };
}

it('sets the virus scan status on pdf files identified by file metadata', async () => {
  const { handler, mocks } = setup();

  const event = makeGuardDutyMalwareScanResultNotificationEvent({
    detail: {
      s3ObjectDetails: {
        bucketName: 'quarantine-bucket',
        objectKey: 'template.pdf',
        versionId: 'pdf-s3-version-id',
      },
      scanResultDetails: { scanResultStatus: 'NO_THREATS_FOUND' },
    },
  });

  mocks.letterUploadRepository.getFileMetadata.mockResolvedValueOnce({
    owner: 'template-owner',
    'template-id': 'template-id',
    'user-filename': 'template.pdf',
    'version-id': 'template-version',
    'file-type': 'pdf-template',
  });

  await handler(event);

  expect(mocks.letterUploadRepository.getFileMetadata).toHaveBeenCalledWith(
    'quarantine-bucket',
    'template.pdf',
    'pdf-s3-version-id'
  );

  expect(
    mocks.templateRepository.setLetterFileVirusScanStatus
  ).toHaveBeenCalledWith(
    { id: 'template-id', owner: 'template-owner' },
    'pdf-template',
    'template-version',
    'PASSED'
  );
});

it('sets the virus scan status on csv files identified by file metadata', async () => {
  const { handler, mocks } = setup();

  const event = makeGuardDutyMalwareScanResultNotificationEvent({
    detail: {
      s3ObjectDetails: {
        bucketName: 'quarantine-bucket',
        objectKey: 'test-data.csv',
        versionId: 'test-data-s3-version-id',
      },
      scanResultDetails: { scanResultStatus: 'NO_THREATS_FOUND' },
    },
  });

  mocks.letterUploadRepository.getFileMetadata.mockResolvedValue({
    owner: 'template-owner',
    'template-id': 'template-id',
    'version-id': 'template-version',
    'file-type': 'test-data',
    'user-filename': 'template.pdf',
  });

  await handler(event);

  expect(mocks.letterUploadRepository.getFileMetadata).toHaveBeenCalledWith(
    'quarantine-bucket',
    'test-data.csv',
    'test-data-s3-version-id'
  );

  expect(
    mocks.templateRepository.setLetterFileVirusScanStatus
  ).toHaveBeenCalledWith(
    { id: 'template-id', owner: 'template-owner' },
    'test-data',
    'template-version',
    'PASSED'
  );
});

it.each($GuardDutyMalwareScanStatusFailed.options)(
  'handles guard duty scan failure status %s',
  async (status) => {
    const { handler, mocks } = setup();

    const event = makeGuardDutyMalwareScanResultNotificationEvent({
      detail: {
        s3ObjectDetails: {
          bucketName: 'quarantine-bucket',
          objectKey: 'template.pdf',
          versionId: 'pdf-s3-version-id',
        },
        scanResultDetails: { scanResultStatus: status },
      },
    });

    mocks.letterUploadRepository.getFileMetadata.mockResolvedValueOnce({
      owner: 'template-owner',
      'template-id': 'template-id',
      'version-id': 'template-version',
      'file-type': 'pdf-template',
      'user-filename': 'template.pdf',
    });

    await handler(event);

    expect(mocks.letterUploadRepository.getFileMetadata).toHaveBeenCalledWith(
      'quarantine-bucket',
      'template.pdf',
      'pdf-s3-version-id'
    );

    expect(
      mocks.templateRepository.setLetterFileVirusScanStatus
    ).toHaveBeenCalledWith(
      { id: 'template-id', owner: 'template-owner' },
      'pdf-template',
      'template-version',
      'FAILED'
    );
  }
);

it('errors if event has no bucket name', async () => {
  const { handler, mocks } = setup();
  const event = {
    detail: {
      s3ObjectDetails: {
        objectKey: 'template.pdf',
        versionId: 'pdf-s3-version-id',
      },
      scanResultDetails: { scanResultStatus: 'NO_THREAT_FOUND' },
    },
  };

  await expect(handler(event)).rejects.toThrowErrorMatchingSnapshot();

  expect(
    mocks.templateRepository.setLetterFileVirusScanStatus
  ).not.toHaveBeenCalled();
});

it('errors if event has no object key name', async () => {
  const { handler, mocks } = setup();
  const event = {
    detail: {
      s3ObjectDetails: {
        bucketName: 'quarantine-bucket',
        versionId: 'pdf-s3-version-id',
      },
      scanResultDetails: { scanResultStatus: 'NO_THREAT_FOUND' },
    },
  };

  await expect(handler(event)).rejects.toThrowErrorMatchingSnapshot();

  expect(
    mocks.templateRepository.setLetterFileVirusScanStatus
  ).not.toHaveBeenCalled();
});

it('errors if event has no object version id', async () => {
  const { handler, mocks } = setup();
  const event = {
    detail: {
      s3ObjectDetails: {
        bucketName: 'quarantine-bucket',
        objectKey: 'template.pdf',
      },
      scanResultDetails: { scanResultStatus: 'NO_THREAT_FOUND' },
    },
  };

  await expect(handler(event)).rejects.toThrowErrorMatchingSnapshot();

  expect(
    mocks.templateRepository.setLetterFileVirusScanStatus
  ).not.toHaveBeenCalled();
});

it('errors if event has no scan result status', async () => {
  const { handler, mocks } = setup();
  const event = {
    detail: {
      s3ObjectDetails: {
        bucketName: 'quarantine-bucket',
        objectKey: 'template.pdf',
        versionId: 'pdf-s3-version-id',
      },
      scanResultDetails: {},
    },
  };

  await expect(handler(event)).rejects.toThrowErrorMatchingSnapshot();

  expect(
    mocks.templateRepository.setLetterFileVirusScanStatus
  ).not.toHaveBeenCalled();
});

it('errors if event has invalid scan result status', async () => {
  const { handler, mocks } = setup();
  const event = {
    detail: {
      s3ObjectDetails: {
        bucketName: 'quarantine-bucket',
        objectKey: 'template.pdf',
        versionId: 'pdf-s3-version-id',
      },
      scanResultDetails: { scanResultStatus: 'UNKNOWN_STATUS' },
    },
  };

  await expect(handler(event)).rejects.toThrowErrorMatchingSnapshot();

  expect(
    mocks.templateRepository.setLetterFileVirusScanStatus
  ).not.toHaveBeenCalled();
});

it('errors if event object metadata cannot be fetched', async () => {
  const { handler, mocks } = setup();
  const event = makeGuardDutyMalwareScanResultNotificationEvent({
    detail: {
      s3ObjectDetails: {
        bucketName: 'quarantine-bucket',
        objectKey: 'template.pdf',
        versionId: 'pdf-s3-version-id',
      },
      scanResultDetails: { scanResultStatus: 'NO_THREATS_FOUND' },
    },
  });

  const error = new Error('GetFileMetadata error');
  mocks.letterUploadRepository.getFileMetadata.mockRejectedValueOnce(error);

  await expect(handler(event)).rejects.toThrow(error);

  expect(
    mocks.templateRepository.setLetterFileVirusScanStatus
  ).not.toHaveBeenCalled();
});

it('errors if status update fails', async () => {
  const { handler, mocks } = setup();
  const event = makeGuardDutyMalwareScanResultNotificationEvent({
    detail: {
      s3ObjectDetails: {
        bucketName: 'quarantine-bucket',
        objectKey: 'template.pdf',
        versionId: 'pdf-s3-version-id',
      },
      scanResultDetails: { scanResultStatus: 'NO_THREATS_FOUND' },
    },
  });

  mocks.letterUploadRepository.getFileMetadata.mockResolvedValueOnce({
    owner: 'template-owner',
    'template-id': 'template-id',
    'version-id': 'template-version',
    'file-type': 'pdf-template',
    'user-filename': 'template.pdf',
  });

  const error = new Error('Status Update error');

  mocks.templateRepository.setLetterFileVirusScanStatus.mockRejectedValueOnce(
    error
  );

  await expect(handler(event)).rejects.toThrow(error);
});
