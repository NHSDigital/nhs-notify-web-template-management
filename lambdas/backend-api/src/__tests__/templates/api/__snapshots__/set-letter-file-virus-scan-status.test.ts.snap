// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`errors if event has invalid scan result status 1`] = `
"[
  {
    "received": "UNKNOWN_STATUS",
    "code": "invalid_enum_value",
    "options": [
      "NO_THREATS_FOUND",
      "THREATS_FOUND",
      "UNSUPPORTED",
      "ACCESS_DENIED",
      "FAILED"
    ],
    "path": [
      "detail",
      "scanResultDetails",
      "scanResultStatus"
    ],
    "message": "Invalid enum value. Expected 'NO_THREATS_FOUND' | 'THREATS_FOUND' | 'UNSUPPORTED' | 'ACCESS_DENIED' | 'FAILED', received 'UNKNOWN_STATUS'"
  }
]"
`;

exports[`errors if event has no bucket name 1`] = `
"[
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "detail",
      "s3ObjectDetails",
      "bucketName"
    ],
    "message": "Required"
  },
  {
    "received": "NO_THREAT_FOUND",
    "code": "invalid_enum_value",
    "options": [
      "NO_THREATS_FOUND",
      "THREATS_FOUND",
      "UNSUPPORTED",
      "ACCESS_DENIED",
      "FAILED"
    ],
    "path": [
      "detail",
      "scanResultDetails",
      "scanResultStatus"
    ],
    "message": "Invalid enum value. Expected 'NO_THREATS_FOUND' | 'THREATS_FOUND' | 'UNSUPPORTED' | 'ACCESS_DENIED' | 'FAILED', received 'NO_THREAT_FOUND'"
  }
]"
`;

exports[`errors if event has no object key name 1`] = `
"[
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "detail",
      "s3ObjectDetails",
      "objectKey"
    ],
    "message": "Required"
  },
  {
    "received": "NO_THREAT_FOUND",
    "code": "invalid_enum_value",
    "options": [
      "NO_THREATS_FOUND",
      "THREATS_FOUND",
      "UNSUPPORTED",
      "ACCESS_DENIED",
      "FAILED"
    ],
    "path": [
      "detail",
      "scanResultDetails",
      "scanResultStatus"
    ],
    "message": "Invalid enum value. Expected 'NO_THREATS_FOUND' | 'THREATS_FOUND' | 'UNSUPPORTED' | 'ACCESS_DENIED' | 'FAILED', received 'NO_THREAT_FOUND'"
  }
]"
`;

exports[`errors if event has no object version id 1`] = `
"[
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "detail",
      "s3ObjectDetails",
      "versionId"
    ],
    "message": "Required"
  },
  {
    "received": "NO_THREAT_FOUND",
    "code": "invalid_enum_value",
    "options": [
      "NO_THREATS_FOUND",
      "THREATS_FOUND",
      "UNSUPPORTED",
      "ACCESS_DENIED",
      "FAILED"
    ],
    "path": [
      "detail",
      "scanResultDetails",
      "scanResultStatus"
    ],
    "message": "Invalid enum value. Expected 'NO_THREATS_FOUND' | 'THREATS_FOUND' | 'UNSUPPORTED' | 'ACCESS_DENIED' | 'FAILED', received 'NO_THREAT_FOUND'"
  }
]"
`;

exports[`errors if event has no scan result status 1`] = `
"[
  {
    "expected": "'NO_THREATS_FOUND' | 'THREATS_FOUND' | 'UNSUPPORTED' | 'ACCESS_DENIED' | 'FAILED'",
    "received": "undefined",
    "code": "invalid_type",
    "path": [
      "detail",
      "scanResultDetails",
      "scanResultStatus"
    ],
    "message": "Required"
  }
]"
`;
